clang/test/SemaSYCL/intel-fpga-local-ast.cpp

```diff
@@ -0,0 +1,514 @@
+
+  //CHECK: VarDecl{{.*}}bank_bits_width
+  //CHECK: SYCLIntelNumBanksAttr{{.*}}Implicit{{$}}
+  //CHECK-NEXT: IntegerLiteral{{.*}}4{{$}}
+  //CHECK-NEXT: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK-NEXT: SYCLIntelBankBitsAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}2
+  //CHECK-NEXT: SubstNonTypeTemplateParmExpr
+  //CHECK-NEXT: NonTypeTemplateParmDecl
+  //CHECK-NEXT: IntegerLiteral{{.*}}2{{$}}
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}3
+  //CHECK-NEXT: IntegerLiteral{{.*}}3{{$}}
+  //CHECK: SYCLIntelBankWidthAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}8
+  //CHECK-NEXT: SubstNonTypeTemplateParmExpr
+  //CHECK-NEXT: NonTypeTemplateParmDecl
+  //CHECK-NEXT: IntegerLiteral{{.*}}8{{$}}
+  [[intel::bank_bits(A, 3), intel::bankwidth(C)]] unsigned int bank_bits_width;
+
+  // Add implicit memory attribute.
+  //CHECK: VarDecl{{.*}}max_replicates
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelMaxReplicatesAttr
+  //CHECK: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}2
+  //CHECK-NEXT: SubstNonTypeTemplateParmExpr
+  //CHECK: IntegerLiteral{{.*}}2{{$}}
+  [[intel::max_replicates(A)]] unsigned int max_replicates;
+
+  // Test that checks template instantiations for different arg values.
+  //CHECK: VarDecl{{.*}}force_p2d_dup
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelForcePow2DepthAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}1
+  //CHECK-NEXT: SubstNonTypeTemplateParmExpr
+  //CHECK-NEXT: NonTypeTemplateParmDecl
+  //CHECK-NEXT: IntegerLiteral{{.*}}1{{$}}
+  [[intel::force_pow2_depth(D)]] unsigned int force_p2d_dup[64];
+}
+
+template <int A>
+struct templ_st {
+  //CHECK: FieldDecl{{.*}}templ_force_p2d_field
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelForcePow2DepthAttr
+  //CHECK: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}0
+  //CHECK-NEXT: SubstNonTypeTemplateParmExpr
+  //CHECK-NEXT: NonTypeTemplateParmDecl
+  //CHECK-NEXT: IntegerLiteral{{.*}}0{{$}}
+  [[intel::force_pow2_depth(A)]] unsigned int templ_force_p2d_field[64];
+};
+
+int main() {
+  deviceQueue.submit([&](sycl::handler &h) {
+    h.single_task<class kernel_function>([]() {
+      check_ast();
+      check_template_parameters<2, 4, 8, 1>();
+      struct templ_st<0> ts {};
+    });
+  });
+
+  return 0;
+}
+
+// Test that checks global constant variable (which allows the redeclaration) since
+// SYCLIntelConstVar is one of the subjects listed for [[intel::max_replicates()]] attribute.
+
+// Check duplicate argument values with implicit memory attribute.
+//CHECK: VarDecl{{.*}}var_max_replicates
+//CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+//CHECK: SYCLIntelMaxReplicatesAttr
+//CHECK-NEXT: ConstantExpr
+//CHECK-NEXT: value:{{.*}}12
+//CHECK-NEXT: IntegerLiteral{{.*}}12{{$}}
+//CHECK: SYCLIntelMaxReplicatesAttr
+//CHECK-NEXT: ConstantExpr
+//CHECK-NEXT: value:{{.*}}12
+//CHECK-NEXT: IntegerLiteral{{.*}}12{{$}}
```

Comment:
Does this need a FIXME and removal of the CHECKs for the duplicate value?
Or is the duplicate value the right behavior?

Same question for the next three as well.